src/box/applier.cc

```diff
@@ -783,12 +783,13 @@ applier_wait_snapshot(struct applier *applier)
 				  (uint32_t) row.type);
 		}
 		/*
-		 * Start vclock. The vclock of the checkpoint
-		 * the master is sending to the replica.
-		 * Used to initialize the replica's initial
-		 * vclock in bootstrap_from_master()
+		 * Start vclock. The vclock of the checkpoint the master is
+		 * sending to the replica. Used to initialize the replica's
+		 * initial vclock in bootstrap_from_master(). Let the journal
+		 * handle this update as a special request to bump vclock.
 		 */
-		xrow_decode_vclock_xc(&row, journal_vclock);
+		if (journal_write_row(&row) != 0)
+			diag_raise();
```

Comment:
I thought about a method for a while. Discarded it because:

- It adds one more function to box.h, which I wanted to avoid. Wanted specifically to use the existing pipeline of vclock update, and keep as few of them as possible.

- One of the reasons for having it updated only in box.cc is protection against people trying to change it outside because in the moment it is convenient, but in the long run it is tech debt. If we have a function which allows to set the vclock, then it won't be different from just having it non-const - people will call this func when they want to change vclock in any moment. It can slip through reviews. But if there is no such function, devs will think more if they actually want to update it outside of box.cc.

Speaking of making this function invocable only once - I am not 100% sure it is set only one time? - I see in applier.cc in `applier_wait_snapshot()` the vclock is expected to be found in a loop when `applier->version_id < version_id(1, 7, 0)`. What if the vclock is actually stored more than once in some older snapshots? I don't remember, but I was quite paranoid.

Reply:
The problem with initializing vclock through `journal_write` is that you create a precedent where applier writes a heartbeat message (IPROTO_OK with a vclock) to journal. Even though this only happens during initial join.

Just like some developer could see vclock update directly from the applier, now he might see that writing IPROTO_OK to the journal is the thing to do, and we'll end up with heartbeats in WAL.

Also I don't like that we start abusing the journal pipeline to set initial instance vclock. This seems strange. Basically bootstrap and initial_join journals were merely stubs, but now initial_join_journal starts to perform some non-obvious task, and only for a specific row type.

> It adds one more function to box.h, which I wanted to avoid. Wanted specifically to use the existing pipeline of vclock update, and keep as few of them as possible.

But why? Adding an `_init` method seems a good addition to the vclock update pipeline. Of course this would only make sense if the function is invokable only once. The place you mention has a `break` after reading vclock, so it seems even `1.6` sets this vclock only once, so I think making sure the vclock is initialized only once is doable.

If you still don't like the idea with `box_init_storage_vclock` let's at least make sure that `initial_join_journal_write` updates the vclock only once and errors when provided with an IPROTO_OK row for the second time.

Reply:
Ok, you are right, I didn't notice that `break` :(. Sure, I introduced the vclock init function.

> Just like some developer could see vclock update directly from the applier, now he might see that writing IPROTO_OK to the journal is the thing to do

The idea was that it is harder to do that just assign global a vclock. That would work as a barrier to let the author think a bit more about his plan.

Reply:
Thanks! I'm happy now.